### ISSUE TITLE
Could not run exla/bench/softmax.exs

### ISSUE THREAD

**Issue by dudeofea** [elixir-nx/nx#372]
I am trying to get setup using EXLA with CUDA, here is my problem:
I first run the docker container like so (same as README):
```
docker run -it \
  -v $PWD:$PWD \
  -e TEST_TMPDIR=$PWD/tmp/bazel_cache \
  -e EXLA_CACHE=$PWD/tmp/exla_cache \
  -e EXLA_FLAGS=--config=cuda \
  -e EXLA_TARGET=cuda \
  -w $PWD \
  --gpus=all \
  --rm exla:cuda10.1 bash
```
Then I run the benchmark (I copied it into my lib folder):
```
mix run lib/softmax.exs
```
And I get the following:
```
#Nx.Tensor<
  f32[1000000]
  [1.5798914319020696e-6, 1.5813845948287053e-6, 6.786568746974808e-7, 7.630458753737912e-7, 9.60705392571981e-7, 1.201145664708747e-6, 7.319711698983156e-7, 8.235413702095684e-7, 1.1704128155543003e-6, 1.3260772675494081e-6, 6.301195867308707e-7, 1.037560764416412e-6, 6.691853400297987e-7, 7.122930014702433e-7, 6.134396812740306e-7, 1.1376204156476888e-6, 5.985975235489605e-7, 6.511505148409924e-7, 1.4426749430640484e-6, 9.475306796957739e-7, 1.4026571761860396e-6, 9.081288112611219e-7, 1.205167450279987e-6, 1.3394895859164535e-6, 1.2977268397662556e-6, 1.215961788147979e-6, 1.3179738971302868e-6, 1.0198649533776916e-6, 7.310582077479921e-7, 6.05523837293731e-7, 1.3892291690353886e-6, 1.0121924560735351e-6, 9.698194389784476e-7, 6.005784030094219e-7, 9.206157756125322e-7, 1.0486402288734098e-6, 1.5657871017538127e-6, 8.14412999261549e-7, 1.331909402324527e-6, 7.479932833120984e-7, 1.0010876394517254e-6, 1.4295559367383248e-6, 7.044977223813476e-7, 1.5564667137368815e-6, 1.3744977422902593e-6, 1.1742151855287375e-6, 1.568202947055397e-6, 8.517512242178782e-7, 8.241113960139046e-7, 6.794938371967874e-7, ...]
>
#Nx.Tensor<
  f32[1000000]
  [1.5798913182152319e-6, 1.5813844811418676e-6, 6.786568746974808e-7, 7.630457616869535e-7, 9.60705392571981e-7, 1.2011454373350716e-6, 7.319711698983156e-7, 8.235413702095684e-7, 1.1704127018674626e-6, 1.3260772675494081e-6, 6.301195298874518e-7, 1.037560764416412e-6, 6.691852831863798e-7, 7.122928877834056e-7, 6.134395675871929e-7, 1.137620301960851e-6, 5.985975235489605e-7, 6.511505148409924e-7, 1.4426748293772107e-6, 9.475305660089361e-7, 1.4026570624992019e-6, 9.081286975742842e-7, 1.2051672229063115e-6, 1.3394895859164535e-6, 1.2977268397662556e-6, 1.215961788147979e-6, 1.317973783443449e-6, 1.0198649533776916e-6, 7.310581509045733e-7, 6.055237804503122e-7, 1.3892289416617132e-6, 1.0121923423866974e-6, 9.698194389784476e-7, 6.00578346166003e-7, 9.206157187691133e-7, 1.0486402288734098e-6, 1.565786988066975e-6, 8.144128287312924e-7, 1.3319092886376893e-6, 7.479931696252606e-7, 1.0010876394517254e-6, 1.429555823051487e-6, 7.044976655379287e-7, 1.5564666000500438e-6, 1.3744977422902593e-6, 1.1742150718418998e-6, 1.5682028333685594e-6, 8.517511673744593e-7, 8.241113960139046e-7, 6.794937803533685e-7, ...]
>

03:16:30.502 [info]  CPU Frequency: 2400000000 Hz

03:16:30.508 [info]  XLA service 0x7f96a00126f0 initialized for platform Host (this does not guarantee that XLA will be used). Devices:

03:16:30.508 [info]    StreamExecutor device (0): Host, Default Version
#Nx.Tensor<
  f32[1000000]
  [1.5798914319020696e-6, 1.5813845948287053e-6, 6.786568746974808e-7, 7.630458753737912e-7, 9.607055062588188e-7, 1.201145664708747e-6, 7.319711698983156e-7, 8.235414270529873e-7, 1.1704128155543003e-6, 1.3260772675494081e-6, 6.301195867308707e-7, 1.037560764416412e-6, 6.691853400297987e-7, 7.122930014702433e-7, 6.134396812740306e-7, 1.1376204156476888e-6, 5.985975235489605e-7, 6.511505148409924e-7, 1.4426749430640484e-6, 9.475306796957739e-7, 1.402657403559715e-6, 9.081288112611219e-7, 1.205167450279987e-6, 1.3394895859164535e-6, 1.2977268397662556e-6, 1.215961788147979e-6, 1.3179740108171245e-6, 1.0198649533776916e-6, 7.310582077479921e-7, 6.05523837293731e-7, 1.3892291690353886e-6, 1.0121924560735351e-6, 9.698194389784476e-7, 6.005784030094219e-7, 9.206157756125322e-7, 1.0486403425602475e-6, 1.5657871017538127e-6, 8.144128855747113e-7, 1.331909402324527e-6, 7.479932833120984e-7, 1.0010876394517254e-6, 1.429555823051487e-6, 7.044977223813476e-7, 1.5564667137368815e-6, 1.3744977422902593e-6, 1.1742151855287375e-6, 1.568202947055397e-6, 8.517513379047159e-7, 8.241113960139046e-7, 6.794938371967874e-7, ...]
>
** (ArgumentError) could not find EXLA client named :cuda, the clients specified in your config files are: [:default]
    (exla 0.1.0-dev) lib/exla/client.ex:27: anonymous fn/1 in EXLA.Client.fetch_client!/1
    (exla 0.1.0-dev) lib/exla/locked_cache.ex:25: EXLA.LockedCache.run/2
    (exla 0.1.0-dev) lib/exla/client.ex:17: EXLA.Client.fetch!/1
    (exla 0.1.0-dev) lib/exla/device_backend.ex:28: EXLA.DeviceBackend.from_binary/3
    lib/softmax.exs:50: (file)
```
But I cannot get the benchmark to run, similarly, I cannot run the `defn` compiler with `{EXLA, client: :cuda}` as it gives the same error. I have compiled EXLA in the docker image (which has the appropriate flags set, I believe), and have tested that NVIDIA works in docker with their nvidia-smi docker image.

I'm not sure how to change my EXLA config so that it will use CUDA, as far as I can tell the library should work, but I'm stuck with this configuration issue.

**Comment by josevalim** [comment]
Each client must be configured. For EXLA in particular, we change the default client to run on CUDA, so you don't need to specify `client: :cuda`:

https://github.com/elixir-nx/nx/blob/main/exla/config/runtime.exs#L5

If you do set client: :cuda, then you need to configure it in the config/runtime.exs file. :)

**Comment by dudeofea** [comment]
Does that mean the benchmark no longer needs `client :cuda`?
https://github.com/elixir-nx/nx/blob/main/exla/bench/softmax.exs#L17

It seems to me that `runtime.exs` uses the `EXLA_TARGET` env variable, which in my case is "cuda" since I'm running it in the docker image. Also, from this line it looks to me like it's using `client: :host`, ie the CPU:
```
03:16:30.508 [info]    StreamExecutor device (0): Host, Default Version
```

I'm pretty new to Elixir, but I'll try seeing if I can find my issue, and post an update.

**Comment by josevalim** [comment]
@dudeofea ah, good catch. Can you please add this to the top of the file:

```elixir
Application.put_env(:exla, :clients,
  default: [platform: :host],
  cuda: [platform: :cuda]
)
```

And if it works you can send a PR. :)

**Comment by seanmor5** [comment]
@dudeofea I have resolved this here: 470f752dc07cc211cedbed97914c73a0d3a701d7

It should run the benchmarks without issue if you have `EXLA_TARGET=cuda` set.
